rfcs/react-components/convergence/component-transitions-on-mount-or-unmount.md

```diff
@@ -0,0 +1,239 @@
+  }
+
+  return <slots.root {...slotProps.root} />;
+};
+```
+
+Styles:
+
+```ts
+import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
+import type { SampleSlots, SampleState } from './Sample.types';
+import type { SlotClassNames } from '@fluentui/react-utilities';
+import { tokens } from '@fluentui/react-theme';
+
+export const SampleClassNames: SlotClassNames<SampleSlots> = {
+  root: 'fui-Sample',
+};
+
+/**
+ * Styles for the root slot
+ */
+const useStyles = makeStyles({
+  root: {
+    opacity: 0,
+    transitionTimingFunction: 'ease-out',
+    transitionProperty: 'opacity',
+    willChange: 'opacity',
+  },
+
+  entering: {
+    transitionDuration: '200ms',
+  },
+
+  exiting: {
+    transitionDuration: '250ms',
+  },
+
+  visible: {
+    opacity: 1,
+  },
+});
+
+export const useSampleStyles_unstable = (state: SampleState): SampleState => {
+  const styles = useStyles();
+
+  state.root.className = mergeClasses(
+    SampleClassNames.root,
+    state.visible && styles.visible,
+    state.entering && styles.entering,
+    state.exiting && styles.exiting,
+    styles.root,
+  );
+
+  return state;
+};
+```
+
+Overriding the transition/animation on the Application side:
+
+```tsx
+import * as React from 'react';
+import { Drawer } from '@fluentui/react-drawer';
+import { makeStyles } from '@fluentui/react-components';
+
+const useStyles = makeStyles({
+  customDuration: {
+    transitionDuration: '500ms',
+  },
+});
+
+export const CustomDuration = () => {
+  const styles = useStyles();
+
+  return <Drawer className={styles.customDuration} />;
+};
+```
+
+#### How it works?
+
+The hook accepts a boolean value that represents the "presence" state. It internally tracks the animation/transition and return boolean values related to its motion state and a ref that needs to be assigned to the element we are tracking.
+
+```ts
+const {
+  /*
+   * Ref used to track the target element that requires animation. It should
+   * be passed to the element responsible for performing the animation.
+   */
+  ref,
+
+  /*
+   * Determines whether the component should be displayed on the screen.
+   *
+   * This property will evaluate to `true` when the provided presence value is true.
+   * Once the provided value changes to false, the hook will monitor the transition/animation
+   * completion and subsequently set this property to false.
+   */
+  shouldRender,
+
+  /*
+   * Indicates whether the component is currently rendered and visible.
+   *
+   * This flag will be set to `true` one frame after the shouldRender value becomes true.
+   * It will be set to `false` when the specified presence value changes to false.
+   */
+  visible,
+
+  /*
+   * Indicates whether the component is currently entering the screen.
+   *
+   * This will be `true` when the element is entering the screen. This happens
+   * one frame after `visible` value becomes true.
+   * It becomes `false` when the transition/animation ends.
+   */
+  entering,
```

Comment:
Does it make sense to have different boolean variables for states? is it possible that the animation state is `entering` and `exiting` at the same time?

IMO would make more sense to have a simple `motionState = 'entering' | 'exited' | 'future state'` and would be a lot more extensible for future changes

Reply:
That is a good idea. I changed this already.